Enable non-breaking extensibility (#240)
openid/OpenID4VP: openid-4-verifiable-presentations-1_0.md

title = "OpenID for Verifiable Presentations - Editor's draft"
abbrev = "openid-4-vp"
ipr = "none"

> workgroup = "OpenID Connect"

Reviewer: It wasn't right before but this isn't right either, is it? Should be dcp?

Reviewer: That's as far as I made it or will make it in this PR this holiday weekend.

Author: OpenID4VP is an OpenID Connect working group specification.  The plan has always been for DCP to do a call for adoption for the specification after the next Implementer's Draft.

As a point of reference, when OpenID4VCI reached Implementer's Draft status after the formation of the DCP working group, [it was adopted by DCP](https://lists.openid.net/pipermail/openid-specs-digital-credentials-protocols/Week-of-Mon-20240415/000259.html).

The last [OpenID4VP Implementer's Draft](https://openid.net/specs/openid-4-verifiable-presentations-1_0-ID2.html) was before [the formation of the DCP working group](https://lists.openid.net/pipermail/openid-specs-digital-credentials-protocols/Week-of-Mon-20230904/000001.html), so this hasn't yet been done for OpenID4VP.

Reviewer: Woah, really? I guess I'd just been assuming that transition had happened after all the hullabaloo around it whenever it happened. And not paying close enough attention. But you know what they say about assuming...

Reviewer: To the extent that all the OIDF procedures around IPR, WGs, etc., actually matter, this seems rather problematic. But what do I know? I'm only on line 5 of one PR...

Author: If you can review the lines after line 5 as well, that would be appreciated.

Reviewer: understood and acknowledging that doing so is on my todo list

Reviewer 2: officially, it is AB/Connect Working Group. not going to die about the string value here, but just saying.

This specification enables the Verifier to send both Presentation Definition JSON object and Client Metadata JSON object by value or by reference.

> This specification defines the following new Authorization Request parameters:

Reviewer 2: ```suggestion
This specification defines the following new request parameters:
```

we have been avoiding to call this request authorization request because this can be a request for any grant type. would prefer we stick to that..

Reviewer 3: this section is called `Authorization Request`(and has plenty more usage of the same language) should that change too?

`response_mode`:
: OPTIONAL. Defined in [@!OAuth.Responses]. This parameter is used (through the new Response Mode `direct_post`) to ask the Wallet to send the response to the Verifier via an HTTPS connection (see (#response_mode_post) for more details). It is also used to request signing and encrypting (see (#jarm) for more details). If the parameter is not present, the default value is `fragment`. 

> Additional Authorization Request parameters MAY be defined and used,

Reviewer 2: ```suggestion
Additional request parameters MAY be defined and used,
```

as described in [@!RFC6749].

> The Wallet MUST ignore any unrecognized parameters.

Reviewer 4: yes, this is also linked to this: https://github.com/openid/OpenID4VP/pull/233/files#r1744570536

The request MUST use the HTTP POST method with the `https` scheme, and the content type `application/x-www-form-urlencoded` and the accept header set to `application/oauth-authz-req+jwt`.

> The following Request URI endpoint parameters are defined:

Reviewer 2: ```suggestion
The following Authorization Request parameters are defined:
```
what is Request URI endpoint parameter...?

Additionally, if the `client_id_scheme` value permits signed Request Objects, the Wallet SHOULD list supported cryptographic algorithms for securing the Request Object through the `request_object_signing_alg_values_supported` parameter. Conversely, the Wallet MUST NOT include this parameter if the `client_id_scheme` precludes signed Request Objects.

> Additional Request URI endpoint parameters MAY be defined and used.

Reviewer 2: ```suggestion
Additional  parameters MAY be defined and used in the request to the Request URI Endpoint.
```

Including the `presentation_submission` parameter as a separate response parameter allows the Wallet to provide the Verifier with additional information about the format and structure in advance of the processing of the VP Token, and can be used even with the Credential formats that do not allow for the direct inclusion of `presentation_submission` parameters inside a Credential itself.

Additional response parameters MAY be defined and used,
as described in [@!RFC6749].

> The client MUST ignore any unrecognized parameters.

Reviewer 4: ```suggestion
The Client MUST ignore any unrecognized parameters.
```

Note: The Verifier's component providing the user interface (Frontend) and the Verifier's component providing the Response URI (Response Endpoint) need to be able to map authorization requests to the respective authorization responses. The Verifier MAY use the `state` Authorization Request parameter to add appropriate data to the Authorization Response for that purpose, for details see (#implementation_considerations_direct_post). 

> Additional Authorization Request parameters MAY be defined and used
> with the Response Mode `direct_post`.
> The Wallet MUST ignore any unrecognized parameters.
> 

Reviewer 2: ```suggestion
Additional request parameters MAY be defined and used with the Response Mode `direct_post`.
The Wallet MUST ignore any unrecognized parameters.

```

Note: In the Response Mode `direct_post` or `direct_post.jwt`, the Wallet can change the UI based on the Verifier's callback to the Wallet following the submission of the Authorization Response.

> Additional parameters for use in the response from the Response Endpoint to the Wallet
> MAY be defined and used.

Reviewer 2: ```suggestion
Additional parameters MAY be defined and used in the response from the Response Endpoint to the Wallet.
```

`client_id_schemes_supported`:
: OPTIONAL. Array of JSON Strings containing the values of the Client Identifier schemes that the Wallet supports. The values defined by this specification are `pre-registered`, `redirect_uri`, `entity_id`, `did`. If omitted, the default value is `pre-registered`. Other values may be used when defined in the profiles of this specification.

Additional wallet metadata parameters MAY be defined and used,
as described in [@!RFC8414].

> The verifier MUST ignore any unrecognized parameters.

Reviewer 4: ```suggestion
The Verifier MUST ignore any unrecognized parameters.
```

## Additional Verifier Metadata Parameters {#client_metadata_parameters}

> This specification defines the following new client metadata parameters according to [@!RFC7591], to be used by the Verifier:

Reviewer 4: ```suggestion
This specification defines the following new Client metadata parameters according to [@!RFC7591], to be used by the Verifier:
```

`vp_formats`:
: REQUIRED. An object defining the formats and proof types of Verifiable Presentations and Verifiable Credentials that a Verifier supports. For specific values that can be used, see (#alternative_credential_formats).
Deployments can extend the formats supported, provided Issuers, Holders and Verifiers all understand the new format.

> Additional verifier metadata parameters MAY be defined and used,

Reviewer 4: ```suggestion
Additional Verifier metadata parameters MAY be defined and used,
```

* `expected_origins`: REQUIRED when signed requests defined in (#signed_request) are used with the W3C Digital Credentials API [@!w3c.digital_credentials_api]. An array of strings, each string representing an origin of the Verifier that is making the request. The Wallet can detect replay of the request from a malicious Verifier by comparing values in this parameter to the origin asserted by the Web Platform.

> Additional Authorization Request parameters MAY be defined and used with this profile,
> as described in [@!RFC6749].
> The Wallet MUST ignore any unrecognized parameters.

Reviewer 2: ```suggestion
Additional request parameters MAY be defined and used with this profile for the W3C Digital Credentials API.
The Wallet MUST ignore any unrecognized parameters.
```

don't think it is authorization request that is used with browser api...

> Additional Authorization Request parameters MAY be defined and used with this profile,
> as described in [@!RFC6749].
> The Wallet MUST ignore any unrecognized parameters.
> 

Reviewer 2: ```suggestion
```
this has already been said, not need to repeat.
